Extend ProcessStartInfo to allow setting LOGON_NETCREDENTIALS_ONLY

hangy opened the pull request:
Fix #28114

hangy committed:
Add test for ProcessStartInfo.UseCredentialsForNetworkingOnly

hangy committed:
Add test to ensure that incompatible LogonFlags cannot be enabled

hangy committed:
Add support for LOGON_NETCREDENTIALS_ONLY

hangy committed:
Replace Theory by Facts

hangy committed:
Add UseCredentialsForNetworkingOnly to ref.cs

hangy commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.Windows.cs:
```diff
@@ -25,6 +25,9 @@ public string Domain
         [SupportedOSPlatform("windows")]
         public bool LoadUserProfile { get; set; }
 
+        [SupportedOSPlatform("windows")]
```
Note that the original issue said to determine the Windows version that the flag first became available in. I haven't found a version history of winbase.h, and [CreateProcessWithLogonW ](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createprocesswithlogonw) only says that the method was added in Windows XP/Windows Server 2003, but doesn't specify if the `LOGON_NETCREDENTIALS_ONLY` logon flag might have been added afterwards. However, [this forum post from 2030](https://forums.codeguru.com/showthread.php?228109-How-to-use-CreateProcessWithLogonW) did already refer to to the `LOGON_NETCREDENTIALS_ONLY` flag. Therefore, based on the OS support of .NET Core, just checking for `"windows"` *should*  be fine.

adamsitnik replied:
Sounds reasonable to me. The oldest Windows we support is 7 and we have CI leg for it. So in case our assumptions turn out to be wrong, the test is going to fail.

hangy commented:
I'm having a bit of an issue constructing a good functional test case for this implementation. I could start a new `Process` with user credentials and `UseCredentialsForNetworkingOnly=true`, and verify that user username of the process doesn't match the name specified in the credentials (because the processes' main user will still be the user running the tests). That'd basically be an invese check of [TestUserCredentialsPropertiesOnWindows](https://github.com/dotnet/runtime/blob/385e1bafaa307736541c6024ab29008f98f400f8/src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs#L511). Would this test be of any use?

To actually test the functionality, we'd need to have some *network* resource that can be accessed with "integrated auth". I skimmed the runtime's codebase a bit to look for other tests that implicitly use authenticated network resources, and didn't find too many. The [Kerberos Testing Environment](https://github.com/dotnet/runtime/blob/84eb453a050324e1e0a3e7643bdab526219d972c/src/libraries/Common/tests/System/Net/Security/Kerberos/README.md) is limited to *nix. One option might be to write a file and read it in the external `Process` via an UNC path, as [ExtractAssociatedIcon_UNCFilePath_Success](https://github.com/dotnet/runtime/blob/6251d496e307ed39f1e600cd036106cc1bd7e340/src/libraries/System.Drawing.Common/tests/IconTests.cs#L378) does, but I'm not sure how involved "external" resources (in this case we'd need set up a network share for the temporary user) are allowed to be in this test category.

hangy committed:
Add WIP TestUserNetworkCredentialsPropertiesOnWindows test

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1355,5 +1374,120 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            string testFileContent = Guid.NewGuid().ToString();
+            File.WriteAllText(testFilePath, testFileContent);
+
+            using WindowsTestAccount accountWithoutFilePermission = new WindowsTestAccount("testForDotnetRuntime");
```
this ctor creates an account with given name: https://github.com/dotnet/runtime/blob/9342649430c19af4bc9fe0d870609778f244a721/src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs#L44

and right after it you are removing the account and creating it again using sys-calls directly.

I really like the idea of `WindowsTestAccount` here (I had no idea that it exists). I think that you should extend it with the possibility to specify user password in the ctor and stop using the sys-calls directly here.

If you don't mind and it's possible, please refactor the `TestUserCredentialsPropertiesOnWindows` test to use the same approach as well:

https://github.com/dotnet/runtime/blob/85fa287fa4dcae54519ea4c4420bba583198df98/src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs#L468

hangy replied:
> this ctor creates an account with given name:
> 
> https://github.com/dotnet/runtime/blob/9342649430c19af4bc9fe0d870609778f244a721/src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs#L44
> 
> and right after it you are removing the account and creating it again using sys-calls directly.

The first test account is `"testForDotnetRuntime"`, and the second is `"testForDotnetNetwork`". It's much clearer in the refactored version with `WindowsIdentityFixture`.

> I really like the idea of `WindowsTestAccount` here (I had no idea that it exists). I think that you should extend it with the possibility to specify user password in the ctor and stop using the sys-calls directly here.

I decided to expose the generated password as a property. Since it's a test helper class, I don't think there's huge risk of leaking secrets here. This way, you don't have to implement password generation in each test that might need it.

I created a separate PR #77804 for the refactoring, and pulled it in here for immediate use.

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1355,5 +1374,120 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            Assert.False(string.IsNullOrWhiteSpace(testFilePath), $"Path to test file should not be empty: {testFilePath}");
+
+            string testFilePathRoot = Path.GetPathRoot(testFilePath);
+            string testFileUncPath = $"\\\\{Environment.MachineName}\\{testFilePath.Substring(0, testFilePathRoot.IndexOf(":"))}$\\{testFilePath.Replace(testFilePathRoot, "")}";
```
using UNC paths to test this feature is a very clever idea! 👍

hangy replied:
I thought so, too, and found this same approach being used in [IconTest](https://github.com/dotnet/runtime/blob/6251d496e307ed39f1e600cd036106cc1bd7e340/src/libraries/System.Drawing.Common/tests/IconTests.cs#L372-L394), and adapted it accordingly. However, since the new account is supposed to be somewhat restricted, it cannot access the default shares like `C$`.
May have to create a new to achieve that, but that's probably still less invasive than hosting any other service that speaks any similar (Kerberos?) auth, which would make use of `LOGON_NETCREDENTIALS_ONLY`.

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1355,5 +1374,120 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+
+            try
+            {
+                // Schedule a process to see what env vars it gets.  Have it write out those variables
+                // to its output stream so we can read them.
```
this seem to be a copy paste error?

```suggestion
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1355,5 +1374,120 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: false);
+                }
+
+                Assert.Equal(Interop.ExitCodes.NERR_Success, Interop.NetUserDel(null, username));
```
it would be great if you could refactor this test and `TestUserCredentialsPropertiesOnWindows` and re-use common logic. It could be something like this:

```cs
private static void CreateUserAndExecute(Process process, string userName, string password)
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1355,5 +1374,120 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+                // to its output stream so we can read them.
+                p = CreateProcess(arg =>
+                {
+                    Console.Write(File.ReadAllText(arg));
```
would it be possible to extend this test with checking that using `testFilePath` to read the file throws?

hangy replied:
Oh, interesting. I actually hadn't thought about that case, but that's a valid point. Will have a look!

adamsitnik commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.Windows.cs:
```diff
@@ -25,6 +25,9 @@ public string Domain
         public bool LoadUserProfile { get; set; }
 
+        [SupportedOSPlatform("windows")]
+        public bool UseCredentialsForNetworkingOnly { get; set; }
```
since this is a new public API, it needs and XML doc

hangy replied:
ACK. I wasn't sure wheter it was okay to simply add them as `///` blocks or if I should add them to `dotnet/dotnet-api-docs` in a separate PR, but now found https://github.com/dotnet/runtime/blob/32ac664fca93cb692597e75021ae5e8ee7b75b1c/docs/coding-guidelines/adding-api-guidelines.md, which answers that question.

adamsitnik reviewed (commented):
It looks very promising, thank you @hangy !

hangy committed:
Allow generated password to be retrieved from WindowsTestAccount

hangy committed:
Use WindowsTestAccount class in TestUserCredentialsPropertiesOnWindows instead of manual Interop

hangy committed:
Merge branch 'refactor-psi-test' into psi-logon-credentials-only

hangy committed:
Remove comment (copy & paste error)

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

hangy committed:
Use WindowsTestAccount for TestUserNetworkCredentialsPropertiesOnWindows

hangy committed:
Test that external process cannot open the file without network credentials

hangy committed:
Add XML docs to UseCredentialsForNetworkingOnly

hangy committed:
fix: Disable UseShellExecute for test

hangy committed:
fix: Provide credentials for test, so that the code is acctually run.

hangy committed:
Merge remote-tracking branch 'origin/main' into psi-logon-credentials-only

hangy committed:
wip: Refactor ProcessStartInfoTests to use common logic for WindowsTestAccount related code

hangy committed:
Merge remote-tracking branch 'origin/main' into psi-logon-credentials-only

hangy committed:
Auto-generate usernames based on the test method's name

adamsitnik commented:
@hangy could you please solve the merge conflicts? I've noticed that you have added some "console" logging. What problem are you facing?

hangy committed:
fix: Explicitly set domain and account in startinfo

hangy committed:
fix: Pass file paths as env paths, because of apparent impats when using args with RemoteExecutor

hangy committed:
test: Reorder order of assertions to make it more obvious

hangy committed:
test: Create temporary share for integration test

Adapted from
https://github.com/dotnet/runtime/blob/b0920912512aa7e205363b046a1911e6a97df31f/src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs#L108-L161

hangy committed:
test: Re-create temporary UNC share if it already exists

hangy commented:
> @hangy could you please solve the merge conflicts? I've noticed that you have added some "console" logging. What problem are you facing?

The console "logging" is just data transfer from the `RemoteExecutor` process as implemented in [TestEnvironmentOfChildProcess()](https://github.com/dotnet/runtime/blob/d099f075e45d2aa6007a22b71b45a08758559f80/src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs#L244). Seems to be the only way to return more than a status code from the executed lambda, because normal variables cannot be marshalled to the process.

The basic "netonly" access seems to work (I just started notepad from the test instead of the `RemoteExecutor` and was able to see the UNC session in `mmc`), but running the `RemoteExecutor` with the temporary impersonated user account is giving me some permission issues. Probably need to figure out some more permissions. 😅

hangy commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,326 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+        }
+    }
+
+    public partial class TestFileShare : IDisposable
```
@adamsitnik I took this code from https://github.com/dotnet/runtime/blob/b0920912512aa7e205363b046a1911e6a97df31f/src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs#L108-L161 and modified it a bit to be a single, disposable file share. Should I move this to a separate class in `Common/tests/TestUtilities` (same level as `WindowsTestAccount`) and refactor `FileStreamConformanceTests` and this test to use it? Or would this increase the scope of this PR too much, since I'm not really touching `System.IO.FileSystem` files for this PR otherwise? Not sure about the usual policies for this repo. :)

adamsitnik replied:
> Should I move this to a separate class in Common/tests/TestUtilities (same level as WindowsTestAccount) and refactor FileStreamConformanceTests and this test to use it?

That would be great! 🥇 

> Or would this increase the scope of this PR too much, since I'm not really touching System.IO.FileSystem files for this PR otherwise? Not sure about the usual policies for this repo. :)

I currently own `System.IO.FileSystem` and I am fine with that ;)

hangy replied:
eef186a90d2cad6e87808ed56f468c5dfeb0f86d

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -467,63 +466,15 @@ public void TestWorkingDirectoryPropertyInChildProcess()
+            using Process p = CreateProcessLong();
+            p.StartInfo.LoadUserProfile = true;
 
+            using var testAccountCleanup = CreateUserAndExecute(p);
```
when type is not obvious we don't use `var` (https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md#c-coding-style)

```suggestion
            using TestProcessState testAccountCleanup = CreateUserAndExecute(p);
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -467,63 +466,15 @@ public void TestWorkingDirectoryPropertyInChildProcess()
+            string username = testAccountCleanup.ProcessAccountName.Split('\\').Last();
+            Assert.Equal(username, Helpers.GetProcessUserName(p));
+            bool isProfileLoaded = GetNamesOfUserProfiles().Any(profile => profile.Equals(username));
+            Assert.True(isProfileLoaded);
```
this refactor has made it much simpler to read and understand the code, great! 👍

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -467,63 +466,15 @@ public void TestWorkingDirectoryPropertyInChildProcess()
-                {
-                    throw new SkipTestException($"{p.StartInfo.FileName} has been locked by some other process");
-                }
+            using Process p = CreateProcessLong();
```
nit: since we started the refactoring, we can use some more self-describing name

```suggestion
            using Process longRunning = CreateProcessLong();
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,326 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            {
+                // Read file content using network credentials and output it for assertion comparison
+                Console.Write(File.ReadAllText(Environment.GetEnvironmentVariable(UncPath)));
+                Console.Write(ItemSeparator);
+
+                try
+                {
+                    // Read file content using current user - should fail, because impersonated user
+                    // isn't explicitly authorized to access the file.
+                    _ = File.ReadAllText(Environment.GetEnvironmentVariable(LocalPath));
+                    Console.Write('1');
+                }
+                catch (SecurityException)
+                {
+                    Console.Write('0');
+                }
+                catch (UnauthorizedAccessException)
+                {
+                    Console.Write('0');
+                }
+
+                return RemoteExecutor.SuccessExitCode;
```
We could simplify it a bit and just use different exit codes to indicate failure rather than writing to standard output:

```suggestion
                try
                {
                    // Read file content using current user - should fail, because impersonated user
                    // isn't explicitly authorized to access the file.
                    _ = File.ReadAllText(Environment.GetEnvironmentVariable(LocalPath));
                    
                    return -1;
                }
                catch (Exception ex) when (ex is SecurityException or UnauthorizedAccessException)
                {
                    return RemoteExecutor.SuccessExitCode;
                }
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,326 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            p.StartInfo.RedirectStandardOutput = true;
+            p.StartInfo.UseCredentialsForNetworkingOnly = true;
+
+            Action<string, string> setup = (username, _) =>
+            {
+                if (PlatformDetection.IsNotWindowsServerCore) // for this particular Windows version it fails with Attempted to perform an unauthorized operation (#46619)
+                {
+                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: true);
+                }
+            };
```
nit: this is very subjective, but you can use local C# functions if you find them easier to use/more readable. We typically put them at the end of the method.

Moreover, you are passing `username` but capturing `testFilePath`, so you can just capture all locals and pass an `Action` instead of `Action<string, string>` to the test utility (normally it is a perf problem, but that is not the case for test project)

```suggestion
            void Setup()
            {
                if (PlatformDetection.IsNotWindowsServerCore) // for this particular Windows version it fails with Attempted to perform an unauthorized operation (#46619)
                {
                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: true);
                }
            };
```

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,326 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+        {
+            string callerIntials = new string(memberName.Where(c => char.IsUpper(c)).Take(18).ToArray());
+
+            WindowsTestAccount processAccount = new WindowsTestAccount(string.Concat("d", callerIntials));
```
this type implements `IDisposable`, when is it going to get disposed?

hangy replied:
It's passed to the ctor of `TestProcessState`. That class is responsible for disposing both accounts, and the caller of `CreateUserAndExecute` is responsible for disposing it.

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,326 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+
+            public string ImpersonationAccountName => _processImpersonationAccount?.AccountName;
+
+            protected virtual void Dispose(bool disposing)
+            {
+                if (!_disposedValue)
+                {
+                    if (disposing)
+                    {
+                        if (_hasStarted)
+                        {
+                            _process.Kill();
+
+                            Assert.True(_process.WaitForExit(WaitInMS));
+                        }
+
+                        if (PlatformDetection.IsNotWindowsServerCore)
+                        {
+                            // remove the access
+                            SetAccessControl((_processImpersonationAccount ?? _processAccount).AccountName, _process.StartInfo.FileName, _workingDirectory, add: false);
+                        }
+
+                        _additionalCleanup?.Invoke(_processAccount?.AccountName, _processImpersonationAccount?.AccountName);
+                        _processAccount?.Dispose();
+                        _processImpersonationAccount?.Dispose();
+                    }
+
+                    _disposedValue = true;
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(disposing: true);
+                GC.SuppressFinalize(this);
+            }
```
this type does not define a finalizer, so there is no need to call `GC.SuppressFinalize(this)`. We also don't expect anyone to derive from it, so there is no need for `virtual void Dispose(bool disposing)`

```suggestion
            public void Dispose()
            {
                if (_hasStarted)
                {
                    _process.Kill();

                    Assert.True(_process.WaitForExit(WaitInMS));
                }

                if (PlatformDetection.IsNotWindowsServerCore)
                {
                    // remove the access
                    SetAccessControl((_processImpersonationAccount ?? _processAccount).AccountName, _process.StartInfo.FileName, _workingDirectory, add: false);
                }

                _additionalCleanup?.Invoke(_processAccount?.AccountName, _processImpersonationAccount?.AccountName);
                _processAccount?.Dispose();
                _processImpersonationAccount?.Dispose();
            }
```

hangy replied:
This seems to be a somewhat subjective topic - some analyzers demand to always use the full disposable pattern with suppressfinalize, etc., while others accept the simplified form. Happy to use whatever you recommend for the project. 👍🏻

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,326 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            }
+        }
+
+        [LibraryImport(Interop.Libraries.Netapi32)]
```
If given test or a utility is Windows specific, we usually make the class `partial`, and move the Windows specific parts to `$fileName.Windows.cs` and modify the project file to build it only for Windows:

https://github.com/dotnet/runtime/blob/9df6ea21007319967975dc9985413bb6518287da/src/libraries/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests.csproj#L44-L47

adamsitnik reviewed (commented):
It looks very promising, but we still need some polishing. Thank you for your contribution @hangy !

hangy committed:
Apply suggestions from code review

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

danmoseley commented:
Some build failures.

hangy committed:
fix: Build errors because of renames

hangy committed:
fix: Make AccountName property setter private

hangy committed:
refactor: Move TestFileShare to TestUtilities

hangy committed:
fix: Use correct assertions

hangy commented on src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs:
```diff
@@ -34,78 +34,82 @@ public sealed partial class WindowsTestAccount : IDisposable
         private SafeAccessTokenHandle _accountTokenHandle;
         public SafeAccessTokenHandle AccountTokenHandle => _accountTokenHandle;
-        public string AccountName { get; private set; }
+        public string AccountName { get; set; }
```
eef186a90d2cad6e87808ed56f468c5dfeb0f86d

hangy commented:
@danmoseley Thanks! Should be fixed in ca186742b85efe65ddfc113526d348b259e35d54

adamsitnik reviewed (commented):
We are very close, thank you for working on this @hangy !

adamsitnik commented:
/azp list

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop-windows

hangy commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,200 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            });
+        }
+
+        [ConditionalFact(nameof(IsAdmin_IsNotNano_RemoteExecutorIsSupported))] // Nano has no "netapi32.dll", Admin rights are required
```
Sounds reasonable. Will update the `TestFileShare` class accordingly.

hangy replied:
Due to the dependency of this method on `ServiceController`, this would require the `TestUtilities` project to take a dependency on the `System.ServiceProcess.ServiceController` library.

What's the *correct* way to do so for the `TestUtilities` project? The `System.IO.FileSystem.Tests` currently uses the `csproj`
```xml
<ProjectReference Include="$(LibrariesProjectRoot)System.ServiceProcess.ServiceController\src\System.ServiceProcess.ServiceController.csproj" />
```
but the `TestUtilities` don't reference library projects in that manner, yet. Using it like this would make the proj compile for the `net6.0` target, but not for the `net462` target. Also, would that require every other `sln` that uses the `TestUtilities` to also reference the `System.ServiceProcess.ServiceController.csproj`?

hangy replied:
Sounds good to me!

hangy commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs:
```diff
@@ -1347,5 +1317,200 @@ private void VerifyNotepadMainWindowTitle(Process process, string filename)
+            p.StartInfo.Environment[UncPath] = testFileUncPath;
+            p.StartInfo.UseCredentialsForNetworkingOnly = true;
+
+            using var processInfo = CreateUserAndExecute(p, true, Setup, Cleanup);
```
I'm getting `"Access is denied."` errors when trying to impersonate. It doesn't *appear* to be related to permissions, because it even happens if I don't use the `RemotExecutor`, but just use a simple process that starts `C:\Windows\System32\notepad.exe`. Maybe there's some minor thing I'm missing?

Basically, this is stopping me from verifying I'm not accidentally reading the file as the current admin user, which would obviously succeed, making the test that the "netonly" user can read the file void.

hangy replied:
> I am not sure if I understand correctly, but your account might not have the permissions to impersonate.

Might be possible, even though I ran the tests as a local admin in an elevated `cmd`/`pwsh`. I was also able to impersonate in `cmd` using `runas`. I don't know if there's a special permission for impersonation, but will have a look. Thanks!

> How about running as admin and extending the tests to verify current user name?

The current user of the `CreateProcess` process probably doesn't change if we use `UseCredentialsForNetworkingOnly`, so it can't be detected. https://stackoverflow.com/a/38390794/11963

danmoseley replied:
> verifying I'm not accidentally reading the file as the current admin user, which would obviously succeed,

I haven't followed this discussion, so this may be random, but an admin can of course modify perms to deny access to itself - if that helps here.

hangy replied:
> > verifying I'm not accidentally reading the file as the current admin user, which would obviously succeed,
> 
> I haven't followed this discussion, so this may be random, but an admin can of course modify perms to deny access to itself - if that helps here.

Definitely! I tried that, too, but it failed for some reason I can't remember, so I ditched the approach at first. I think it's the one of the least intrusive ways to write this test so, and so think I'll have another look at this, too. Thanks for jumping in, @danmoseley!

danmoseley replied:
https://github.com/dotnet/runtime/blob/cb281adedfef0ab34aba947320385f3999f9f82f/src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs#L23

hangy replied:
Thanks! Will test that in the coming days. (It's getting late :wink:)

hangy committed:
refactor: WindowsTestFileShare improvements

- Move TestFileShare to WindowsTestFileShare
- Move ConditionalAttribute logic to WindowsTestFileShare
- Apply condition of WindowsTestFileShare to ProcessStartInfoTests

hangy committed:
refactor: Simplify if-condition

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>

hangy committed:
refactor: Remove impersonation from CreateUserAndExecute, and move setup/cleanup to specific places

hangy committed:
chg: Forbid current user to read the testFile

hangy committed:
Move TestUserNetworkCredentialsPropertiesOnWindows to Windows specific test file to avoid build errors on other platforms

hangy committed:
test: Try not setting too many file permissions explicitly

hangy commented:
So, I've tested this quite a bit with denied permissions, etc., and have come to the conclusion that the feature itself works, but perhaps cannot be tested locally easily.

I found out that even if I don't allow the created user to access my temporary files at all, the `RemoteExecutor` process is able to access the file and the test succeeds. To figure that out, I modified the code run in the RemoteExecutor to keep the file open a bit longer to check access to the file share, and file permissions:
1. I can see that the UNC share is accessed by the temp user ![Screenshot of mmc showing the temporary user accessing the temporary file](https://user-images.githubusercontent.com/87124/203172715-70ae857e-67bc-4717-a22c-07eece342276.png)
2. Even though the user has no effective permissions to the file (neither via UNC or direkt file access), it can access the file: ![Screenshot of the temporary user's effective permissions to the file](https://user-images.githubusercontent.com/87124/203174104-658523da-9ec5-47ea-9137-8b661e431782.jpg)
3. The test still passes.

Additionally, I tried something different: I set up a simple ASP.NET REST API with Negotiate auth on the same PC that runs the test, had it return the current user's username, and called the API from the test: The service returned my username instead of the temporary user's name.

I'm either missing something really basic, or maybe this `LOGON_NETCREDENTIALS_ONLY` flag really only works well on actual remote network sources? I don't have a good source to test it with on my home network, unfortunately. Even if helps, I see no way to write the test in that case.

(Sorry for the German screens, but I don't have any other language packs installed here, right now.)

danmoseley commented:
@bartonjs can you help explain what @hangy is seeing above?

adamsitnik commented:
I am going to reopen this PR as I believe that it adds value and test imperfections should not block us from adding feature support.

adamsitnik committed:
Merge branch 'main' into psi-logon-credentials-only

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.Windows.cs:
```diff
@@ -0,0 +1,78 @@
+                {
+                    // Read file content using network credentials and output it for assertion comparison
+                    _ = File.ReadAllText(Environment.GetEnvironmentVariable(UncPath));
+                    return RemoteExecutor.SuccessExitCode;
+                }
+                catch (Exception ex) when (ex is SecurityException or UnauthorizedAccessException)
+                {
+                    return -1;
+                }
+            });
+            p.StartInfo.Environment[LocalPath] = testFilePath;
+            p.StartInfo.Environment[UncPath] = testFileUncPath;
+            p.StartInfo.UseCredentialsForNetworkingOnly = true;
+
+            using var processInfo = CreateUserAndExecute(p, Setup, Cleanup);
+
+            string processUserName = Helpers.GetProcessUserName(p);
+            Assert.True(p.WaitForExit(WaitInMS));
+
+            Assert.Equal(expectedExitCode, p.ExitCode);
+            Assert.Equal(Environment.UserName, processUserName);
+
+            void Setup(string username, string _)
+            {
+                if (authorizeUserToAccessTestFile && PlatformDetection.IsNotWindowsServerCore) // for this particular Windows version it fails with Attempted to perform an unauthorized operation (#46619)
+                {
+                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: true);
+                }
+            }
+
+            void Cleanup(string username, string _)
+            {
+                if (authorizeUserToAccessTestFile && PlatformDetection.IsNotWindowsServerCore)
+                {
+                    // remove the access
+                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: false);
+                }
+            }
+        }
+    }
```
We can just remove test case which can't be easily tested

```suggestion
        [ConditionalFact(nameof(IsAdmin_IsNotNano_RemoteExecutorIsSupported_CanShareFiles))] // Nano has no "netapi32.dll", Admin rights are required
        [PlatformSpecific(TestPlatforms.Windows)]
        [OuterLoop("Requires admin privileges")]
        public void TestUserNetworkCredentialsPropertiesOnWindows()
        {
            const string ShareName = "testForDotNet";
            const string TestFileContent = "42";
            const string UncPathEnvVar = nameof(UncPathEnvVar);

            string testFilePath = GetTestFilePath();
            File.WriteAllText(testFilePath, TestFileContent);

            using WindowsTestFileShare fileShare = new WindowsTestFileShare(ShareName, Path.GetDirectoryName(testFilePath));
            string testFileUncPath = $"\\\\{Environment.MachineName}\\{ShareName}\\{Path.GetFileName(testFilePath)}";

            using Process process = CreateProcess(() =>
            {
                try
                {
                    Assert.Equal(TestFileContent, File.ReadAllText(Environment.GetEnvironmentVariable(UncPathEnvVar)));

                    return RemoteExecutor.SuccessExitCode;
                }
                catch (Exception ex) when (ex is SecurityException or UnauthorizedAccessException)
                {
                    return -1;
                }
            });
            process.StartInfo.Environment[UncPathEnvVar] = testFileUncPath;
            process.StartInfo.UseCredentialsForNetworkingOnly = true;

            using TestProcessState processInfo = CreateUserAndExecute(process, Setup, Cleanup);

            Assert.Equal(Environment.UserName, Helpers.GetProcessUserName(process));

            Assert.True(process.WaitForExit(WaitInMS));
            Assert.Equal(RemoteExecutor.SuccessExitCode, process.ExitCode);

            void Setup(string username, string _)
            {
                if (PlatformDetection.IsNotWindowsServerCore) // for this particular Windows version it fails with Attempted to perform an unauthorized operation (#46619)
                {
                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: true);
                }
            }

            void Cleanup(string username, string _)
            {
                if (PlatformDetection.IsNotWindowsServerCore)
                {
                    // remove the access
                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: false);
                }
            }
        }
```

adamsitnik commented on src/libraries/System.IO.FileSystem/tests/System.IO.FileSystem.Tests.csproj:
```diff
@@ -234,6 +234,7 @@
     <Compile Include="$(CommonPath)System\IO\PathInternal.CaseSensitivity.cs" Link="Common\System\IO\PathInternal.CaseSensitivity.cs" />
     <Compile Include="$(CommonTestPath)System\IO\ReparsePointUtilities.cs" Link="Common\System\IO\ReparsePointUtilities.cs" />
     <Compile Include="$(CommonTestPath)TestUtilities\System\DisableParallelization.cs" Link="Common\TestUtilities\System\DisableParallelization.cs" />
+    <Compile Include="$(CommonTestPath)TestUtilities\System\WindowsTestFileShare.cs" Link="Common\TestUtilities\System\WindowsTestFileShare.cs" />
```
this needs to be included only for Windows. Since I can't push to your fork I need to be "creative" with the suggestions so I can apply them from GH web UI.

```suggestion
  </ItemGroup>
  <ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'windows'">
    <Compile Include="$(CommonTestPath)TestUtilities\System\WindowsTestFileShare.cs" Link="Common\TestUtilities\System\WindowsTestFileShare.cs" />
```

adamsitnik reviewed (commented):
Big thanks for your contribution @hangy!

adamsitnik committed:
Apply suggestions from code review

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop-windows

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessStartInfoTests.Windows.cs:
```diff
@@ -0,0 +1,71 @@
+                    SetAccessControl(username, testFilePath, Path.GetDirectoryName(testFilePath), add: false);
+                }
+            }
+        }
+}
```
```suggestion
        }
    }
}
```

adamsitnik commented on src/libraries/Common/tests/TestUtilities/System/WindowsTestFileShare.cs:
```diff
@@ -0,0 +1,118 @@
+    {
+        private static readonly Lazy<bool> _canShareFiles = new Lazy<bool>(() =>
+        {
+            if (!PlatformDetection.IsWindowsAndElevated)
```
```suggestion
            if (!PlatformDetection.IsPrivilegedProcess)
```

adamsitnik committed:
Apply suggestions from code review

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop-windows

adamsitnik reviewed (approved):
@hangy once again thank you for your contribution!

adamsitnik commented:
Test failures are unrelated, merging

hangy commented:
> @hangy once again thank you for your contribution!

Thanks for your continued input, and reviews, @adamsitnik!
